node/db/src/rolling/gc.rs

```diff
@@ -0,0 +1,180 @@
+//! 3. delete `old` database(s)
+//! 4. sets `current` database to a newly created one
+//!
+//! ## Correctness
+//! This algorithm considers all blocks that are visited during the snapshot
+//! export task reachable, and ensures they are all transferred and kept in the
+//! current DB space. A snapshot can be used to bootstrap a node from
+//! scratch thus the algorithm is considered appropriate when the post-GC
+//! database contains blocks that are sufficient for exporting a snapshot
+//!
+//! ## Disk usage
+//! During `walk_snapshot`, data from the `old` DB is duplicated in the
+//! `current` DB, which uses extra disk space of up-to 100% of the snapshot file
+//! size
+//!
+//! ## Memory usage
+//! During the data carry-over process, a memory buffer with a fixed capacity is
+//! used to speed up the database write operation
+//!
+//! ## Scheduling
+//! 1. GC is triggered automatically when current DB size is greater than 50% of
+//! the old DB size
+//! 2. GC can be triggered manually by `forest-cli db gc` command
+//! 3. There's global GC lock to ensure at most one GC job is running
+
+use std::time::Duration;
+
+use chrono::Utc;
+use forest_blocks::Tipset;
+use forest_ipld::util::*;
+use fvm_ipld_blockstore::Blockstore;
+use tokio::sync::Mutex;
+
+use super::*;
+use crate::{Store, StoreExt};
+
+pub struct DbGarbageCollector<F>
+where
+    F: Fn() -> Tipset + Send + Sync + 'static,
+{
+    db: RollingDB,
+    get_tipset: F,
+    lock: Mutex<()>,
+    gc_tx: flume::Sender<flume::Sender<anyhow::Result<()>>>,
+    gc_rx: flume::Receiver<flume::Sender<anyhow::Result<()>>>,
+}
+
+impl<F> DbGarbageCollector<F>
+where
+    F: Fn() -> Tipset + Send + Sync + 'static,
+{
+    pub fn new(db: RollingDB, get_tipset: F) -> Self {
+        let (gc_tx, gc_rx) = flume::unbounded();
+
+        Self {
+            db,
+            get_tipset,
+            lock: Default::default(),
+            gc_tx,
+            gc_rx,
+        }
+    }
+
+    pub fn get_tx(&self) -> flume::Sender<flume::Sender<anyhow::Result<()>>> {
+        self.gc_tx.clone()
+    }
+
+    pub async fn collect_loop_passive(&self) -> anyhow::Result<()> {
+        loop {
+            // Check every 10 mins
+            tokio::time::sleep(Duration::from_secs(10 * 60)).await;
+
+            // Bypass size checking during import
+            let tipset = (self.get_tipset)();
+            if tipset.epoch() == 0 {
+                continue;
+            }
+
+            // Bypass size checking when lock is held
+            {
+                let lock = self.lock.try_lock();
+                if lock.is_err() {
+                    continue;
+                }
+            }
+
+            if let (Ok(total_size), Ok(current_size)) = (
+                self.db.total_size_in_bytes(),
+                self.db.current_size_in_bytes(),
+            ) {
+                // Collect when size of young partition > 0.5 * size of old partition
```

Comment:
I'm not convinced by this metric for when to trigger a garbage collection. It's quite difficult to reason about the size of the disk space required in relation to the live data.

How about we track how much live data there was at the last collection and then base the calculation around that number?

Reply:
Any suggestions on what formula to use? I use 0.5 because if the first GC is triggered when `size(current)` >= `size(old)`, the live data size at GC ends will be `size(current)` + `size(new data during GC)` the threshold will be ever-growing

Reply:
P.S. As I understand, `size(old)` is the live data size after previous GC, let me know if u were refering to sth else

Reply:
`size(old)` is not the size of the reachable data after a GC run. When a GC run is started, both the `old` and `new` partition will contain some unreachable data. Therefore, after a GC run has finished, the `old` partition will contain both reachable and unreachable data.

Reply:
Ok I can get the exact size of the reachable data, any suggestion on how to decide the threshold with it?

Reply:
Maybe when the size of the total database is twice the reachable data.

Reply:
Done